File: examples/python_library/player_board.py
from __future__ import annotations

import argparse
import sys
import time
from typing import Any

from rlstats_example import RLStatsStream


def pick(mapping: dict[str, Any], *keys: str, default: Any = None) -> Any:
    for key in keys:
        if key in mapping:
            return mapping[key]
    return default


def to_int(value: Any, fallback: int = -1) -> int:
    if isinstance(value, bool):
        return int(value)
    if isinstance(value, (int, float)):
        return int(value)
    return fallback


def fmt_int(value: Any) -> str:
    if isinstance(value, bool):
        return str(int(value))
    if isinstance(value, (int, float)):
        return str(int(value))
    return "-"


def fmt_speed(value: Any) -> str:
    if isinstance(value, (int, float)):
        return f"{value:.0f}"
    return "-"


def short_primary_id(value: Any) -> str:
    if not isinstance(value, str) or not value:
        return "-"
    return value.split("|")[-1]


def team_score(teams: list[dict[str, Any]], team_num: int) -> int:
    for team in teams:
        if to_int(pick(team, "TeamNum", "team_num"), fallback=-1) == team_num:
            return to_int(pick(team, "Score", "score"), fallback=-1)
    return -1


def render_board(data: dict[str, Any]) -> None:
    players = pick(data, "Players", "players", default=[])
    game = pick(data, "Game", "game", default={})

    if not isinstance(players, list):
        players = []
    if not isinstance(game, dict):
        game = {}

    teams = pick(game, "Teams", "teams", default=[])
    if not isinstance(teams, list):
        teams = []

    blue = team_score(teams, 0)
    orange = team_score(teams, 1)
    frame = to_int(pick(game, "Frame", "frame"), fallback=-1)
    time_seconds = to_int(
        pick(game, "TimeSeconds", "time_seconds"),
        fallback=-1,
    )
    match_guid = pick(data, "MatchGuid", "match_guid", default="-")

    table_rows: list[dict[str, Any]] = []
    for raw_player in players:
        if isinstance(raw_player, dict):
            table_rows.append(raw_player)

    table_rows.sort(
        key=lambda player: (
            to_int(pick(player, "TeamNum", "team_num"), fallback=-1),
            str(pick(player, "Name", "name", default="")).lower(),
        )
    )

    sys.stdout.write("\x1b[2J\x1b[H")
    sys.stdout.write(
        f"Match={match_guid}  Frame={frame}  Time={time_seconds}s  "
        f"Score={blue}-{orange}  Players={len(table_rows)}\n"
    )
    sys.stdout.write(
        f"{'Team':<4} {'Player':<18} {'Primary':<7} {'Score':>6} {'G':>5} {'A':>5} "
        f"{'S':>5} {'Shots':>6} {'Touch':>6} {'Boost':>6} {'Speed':>7}\n"
    )
    sys.stdout.write(
        f"{'-' * 4:<4} {'-' * 18:<18} {'-' * 7:<7} {'-' * 6:>6} {'-' * 5:>5} {'-' * 5:>5} "
        f"{'-' * 5:>5} {'-' * 6:>6} {'-' * 6:>6} {'-' * 6:>6} {'-' * 7:>7}\n"
    )

    for player in table_rows:
        sys.stdout.write(
            f"{fmt_int(pick(player, 'TeamNum', 'team_num')):<4} "
            f"{str(pick(player, 'Name', 'name', default='-')):<18} "
            f"{short_primary_id(pick(player, 'PrimaryId', 'primary_id')):<7} "
            f"{fmt_int(pick(player, 'Score', 'score')):>6} "
            f"{fmt_int(pick(player, 'Goals', 'goals')):>5} "
            f"{fmt_int(pick(player, 'Assists', 'assists')):>5} "
            f"{fmt_int(pick(player, 'Saves', 'saves')):>5} "
            f"{fmt_int(pick(player, 'Shots', 'shots')):>6} "
            f"{fmt_int(pick(player, 'Touches', 'touches')):>6} "
            f"{fmt_int(pick(player, 'Boost', 'boost')):>6} "
            f"{fmt_speed(pick(player, 'Speed', 'speed')):>7}\n"
        )

    sys.stdout.flush()


def parse_args() -> argparse.Namespace:
    parser = argparse.ArgumentParser(
        description=(
            "Render all players in-place from UpdateState events (no scrolling spam)."
        )
    )
    parser.add_argument("--host", default="127.0.0.1")
    parser.add_argument("--port", type=int, default=49123)
    parser.add_argument("--ini-path")
    parser.add_argument(
        "--refresh-ms",
        type=int,
        default=200,
        help="Minimum milliseconds between redraws",
    )
    return parser.parse_args()


def main() -> None:
    args = parse_args()

    stream = RLStatsStream(
        host=args.host,
        port=args.port,
        ini_path=args.ini_path,
    )
    stream.connect()

    last_render = 0.0
    refresh_seconds = max(args.refresh_ms, 0) / 1000.0

    try:
        for event in stream.iter_filtered_events(event_types=["update_state"]):
            now = time.monotonic()
            if now - last_render < refresh_seconds:
                continue

            render_board(event.data)
            last_render = now
    except KeyboardInterrupt:
        pass
    finally:
        stream.close()
        print("\nplayer board stopped")


if __name__ == "__main__":
    main()

File: src/bin/player_board.rs
use std::io::{self, Write};
use std::path::PathBuf;
use std::time::{Duration, Instant};

use rlstatsapi::{ClientOptions, RocketLeagueStatsClient, StatsEvent};

#[derive(Debug, Clone)]
struct CliOptions {
    ini_path: Option<PathBuf>,
    host: Option<String>,
    port: Option<u16>,
    refresh_ms: u64,
}

#[tokio::main]
async fn main() -> Result<(), Box<dyn std::error::Error>> {
    let cli = parse_args()?;

    let mut options = ClientOptions::default();
    if let Some(path) = cli.ini_path.clone() {
        options.stats_api_ini_path = Some(path);
    } else {
        options.auto_enable_packet_rate = false;
    }

    if let Some(host) = cli.host.clone() {
        options.host = host;
    }
    if let Some(port) = cli.port {
        options.port_override = Some(port);
    }

    let mut client = RocketLeagueStatsClient::connect_with_retry(
        options,
        20,
        Duration::from_millis(250),
    )
    .await?;

    let mut last_render =
        Instant::now() - Duration::from_millis(cli.refresh_ms);

    println!(
        "Tracking players from {} (refresh={}ms)",
        client.connection().socket_address(),
        cli.refresh_ms
    );

    loop {
        match client.next_event().await {
            Ok(Some(StatsEvent::UpdateState(state))) => {
                if last_render.elapsed()
                    >= Duration::from_millis(cli.refresh_ms)
                {
                    render_state(&state)?;
                    last_render = Instant::now();
                }
            }
            Ok(Some(_)) => {}
            Ok(None) => {
                eprintln!("connection closed; attempting reconnect...");
                reconnect_with_backoff(&mut client).await?;
            }
            Err(error) => {
                eprintln!("stream error: {error}; attempting reconnect...");
                reconnect_with_backoff(&mut client).await?;
            }
        }
    }
}

async fn reconnect_with_backoff(
    client: &mut RocketLeagueStatsClient,
) -> Result<(), Box<dyn std::error::Error>> {
    let mut last_error: Option<Box<dyn std::error::Error>> = None;

    for _ in 0..10 {
        match client.reconnect().await {
            Ok(()) => return Ok(()),
            Err(error) => {
                last_error = Some(Box::new(error));
                tokio::time::sleep(Duration::from_millis(500)).await;
            }
        }
    }

    match last_error {
        Some(error) => Err(error),
        None => Err("reconnect failed".into()),
    }
}

fn render_state(
    data: &rlstatsapi::events::UpdateStateData,
) -> Result<(), Box<dyn std::error::Error>> {
    let mut players = data.players.iter().collect::<Vec<_>>();
    players.sort_by(|left, right| {
        let left_team = left.team_num.unwrap_or(-1);
        let right_team = right.team_num.unwrap_or(-1);
        left_team.cmp(&right_team).then_with(|| {
            left.name
                .as_deref()
                .unwrap_or("")
                .cmp(right.name.as_deref().unwrap_or(""))
        })
    });

    let blue = team_score(data, 0).unwrap_or(-1);
    let orange = team_score(data, 1).unwrap_or(-1);

    let mut out = io::stdout().lock();
    write!(out, "\x1b[2J\x1b[H")?;

    writeln!(
        out,
        "Match={}  Frame={}  Time={}s  Score={}-{}  Players={}",
        data.match_guid.as_deref().unwrap_or("-"),
        data.game.frame.unwrap_or(-1),
        data.game.time_seconds.unwrap_or(-1),
        blue,
        orange,
        players.len()
    )?;
    writeln!(
        out,
        "{:<4} {:<18} {:<7} {:>6} {:>5} {:>5} {:>5} {:>6} {:>6} {:>6} {:>7}",
        "Team",
        "Player",
        "Primary",
        "Score",
        "G",
        "A",
        "S",
        "Shots",
        "Touch",
        "Boost",
        "Speed"
    )?;
    writeln!(
        out,
        "{:-<4} {:-<18} {:-<7} {:-<6} {:-<5} {:-<5} {:-<5} {:-<6} {:-<6} {:-<6} {:-<7}",
        "", "", "", "", "", "", "", "", "", "", ""
    )?;

    for player in players {
        let primary_short = player
            .primary_id
            .as_deref()
            .map(short_primary_id)
            .unwrap_or("-");
        writeln!(
            out,
            "{:<4} {:<18} {:<7} {:>6} {:>5} {:>5} {:>5} {:>6} {:>6} {:>6} {:>7}",
            fmt_opt_i64(player.team_num),
            player.name.as_deref().unwrap_or("-"),
            primary_short,
            fmt_opt_i64(player.score),
            fmt_opt_i64(player.goals),
            fmt_opt_i64(player.assists),
            fmt_opt_i64(player.saves),
            fmt_opt_i64(player.shots),
            fmt_opt_i64(player.touches),
            fmt_opt_i64(player.effective_boost()),
            fmt_opt_f64(player.effective_speed())
        )?;
    }

    out.flush()?;
    Ok(())
}

fn short_primary_id(value: &str) -> &str {
    value.split('|').next_back().unwrap_or(value)
}

fn fmt_opt_i64(value: Option<i64>) -> String {
    value
        .map(|v| v.to_string())
        .unwrap_or_else(|| "-".to_string())
}

fn fmt_opt_f64(value: Option<f64>) -> String {
    value
        .map(|v| format!("{v:.0}"))
        .unwrap_or_else(|| "-".to_string())
}

fn team_score(
    data: &rlstatsapi::events::UpdateStateData,
    team_num: i64,
) -> Option<i64> {
    data.game
        .teams
        .iter()
        .find(|team| team.team_num == Some(team_num))
        .and_then(|team| team.score)
}

fn parse_args() -> Result<CliOptions, Box<dyn std::error::Error>> {
    let mut args = std::env::args().skip(1);

    let mut ini_path = None;
    let mut host = None;
    let mut port = None;
    let mut refresh_ms = 200u64;

    while let Some(arg) = args.next() {
        match arg.as_str() {
            "--ini" => {
                let value = args.next().ok_or("--ini requires a file path")?;
                ini_path = Some(PathBuf::from(value));
            }
            "--host" => {
                let value = args.next().ok_or("--host requires a value")?;
                host = Some(value);
            }
            "--port" => {
                let value = args.next().ok_or("--port requires a value")?;
                port = Some(value.parse::<u16>()?);
            }
            "--refresh-ms" => {
                let value =
                    args.next().ok_or("--refresh-ms requires a value")?;
                refresh_ms = value.parse::<u64>()?;
            }
            "-h" | "--help" => {
                print_help();
                std::process::exit(0);
            }
            other => {
                return Err(format!("Unknown argument: {other}").into());
            }
        }
    }

    Ok(CliOptions {
        ini_path,
        host,
        port,
        refresh_ms,
    })
}

fn print_help() {
    println!(
        "Usage: player_board [--ini <path>] [--host <host>] [--port <port>] [--refresh-ms <n>]\n\nContinuously renders all players in-place using UpdateState events (no scrolling spam).\n\nExamples:\n  player_board\n  player_board --refresh-ms 100\n  player_board --ini /path/to/DefaultStatsAPI.ini"
    );
}
